KIPs/kip-247.md

```diff
@@ -0,0 +1,186 @@
+- L2: `LendTx.from` is the proposer itself.
+- L3: `LendTx.to` is `GaslessSwapTx.from`
+- L4: `LendTx.value` is `amountLend`
+  - `amountLend` is `R2 + R3` (from aforementioned `amountRepay` definition).
+
+### GaslessSwapRouter
+
+GaslessSwapRouter is a smart contract responsible for token swap and repayment. It must support the following interface:
+
+```
+interface IKIP247 {
+    function swapForGas(address token, uint256 amountIn, uint256 minAmountOut, uint256 amountRepay) external;
+    function addToken(address token, address dex) external;
+    function removeToken(address token) external;
+
+    function claimCommission() external;
+    function updateCommissionRate(uint256 _commissionRate) external;
+
+    // view functions
+    function dexAddress(address token) external view returns (address dex);
+    function getAmountIn(address token, uint amountOut) external view returns (uint amountIn);
+}
+```
+
+`swapForGas()` must ensure that all the following conditions are satisfied:
+
+- R1: Sender has enough tokens. That is, `token.balanceOf(msg.sender) >= amountIn`.
+- R2: Token is whitelisted and has a corresponding DEX address other than zero.
+- R3: `amountReceived >= minAmountOut >= amountRepay` where `amountReceived` is the swap output.
+
+`swapForGas()` must distribute the swap output (`amountReceived`) as follows:
+
+- The block proposer (`block.coinbase`) receives `amountRepay`.
+- User receives `(amountReceived - amountRepay) * (1 - commissionRate)`.
+- The commission `(amountReceived - amountRepay) * commissionRate` is stored in the contract for later claim.
+
+### TxPool Policy Updates
+
+#### GaslessTx Validation (validateTx)
+
+To prevent GaslessTx to be discarded, the validation logic at `txpool.queue` insertion needs to detect GaslessTx and skip the account balance check.
+
+```
+// txpool.validateTx()
+if ...
+else: // if tx is NOT feeDelegatedTx
+    if senderBalance.Cmp(tx.Cost()) < 0:
+        if tx is GaslessTx:
+            validation ok
+        else:
+            insufficientFund error
+```
+
+#### Promoting GaslessTx (promoteExecutables)
+
+GaslessApproveTx cannot be executed unless the corresponding GaslessSwapTx from the same sender is already in the transaction pool.
+Conversely, if the token has not been approved, GaslessSwapTx cannot be executed before GaslessApproveTx.
+In such cases, both transactions need to be promoted simultaneously.
+
+```
+// txpool.promoteExecutables()
+if tx is GaslessApproveTx:
+    if GaslessSwapTx is in queue:
+        promote GaslessApproveTx and GaslessSwapTx to pending
+    else:
+        push GaslessApproveTx to queue
+
+if tx is GaslessSwapTx:
+    if tx is GaslessApproveTx is in queue:
+        promote GaslessApproveTx and GaslessSwapTx to pending
+    else if GaslessSwapTx.nonce == getNonce(user):
+        push GaslessApproveTx to pending
+    else if GaslessSwapTx.nonce == getNonce(user) + 1:
+        promote GaslessSwapTx to queue
```

Comment:
You mixed up promote to pending and push to queue

Reply:
fixed 184d4ea